web/html/src/manager/virtualization/guests/console/guests-console.js

```diff
@@ -53,6 +59,10 @@ class GuestsConsole extends React.Component<Props, State> {
     this.pageUnloading = false;
     this.openVirtSocket();
 
+    // Schedule token refresh a minute before its expiration
+    const tokenExp = getTokenExpiration(props.token);
+    setTimeout(this.refreshToken, tokenExp - Date.now() - 60000);
```

Comment:
Leaving timeouts lingering like this can lead to heisenbugs that are hard to track down, instead I would recommend turning this into a periodic check and handling one reference to the interval. Perhaps something along these lines?

```ts
class GuestsConsole extends React.Component<Props, State> {
  intervalId?: number;
  // This is intentionally not in state since we don't want this alone to trigger a rerender
  isRefreshing = false;

  componentDidMount() {
    this.intervalId = window.setInterval(() => this.refreshToken(), ...);
  }

  componentWillUnmount() {
    if (this.intervalId) {
      window.clearInterval(this.intervalId);
    }
  }

  async refreshToken() {
    if (this.isRefreshing) {
      return;
    }
    this.isRefreshing = true;
    try {
      const response = await Network.post(...).promise;
    } catch(e) {
      // ...
    }
    this.isRefreshing = false;
  }
}
```

Reply:
I moved the interval init / destroy in the component constructor and window close handler since I feared the `componentDidMount()` could be called at each render. Is that OK?

Reply:
I _think_ this is not entirely correct but I'm not quite sure. I don't know how `window.onbeforeunload` works with the hybrid SPA setup that we have, I'm not sure whether it fires when you move between two SPA pages for example. Perhaps @ncounter knows? As for your question, `componentDidMount()` is be fired only once when the component is created, not on each render. Likewise `componentWillUnmount()` will be fire once when the component is being destroyed.

Reply:
From what I know with the SPA engine we swap the section content (the monitored frame defined in the SPA configuration) once we get the new data, but there is no `unloading` of resources in the `window` object. The `window.onbeforeunload` event is not touched at all, it is used in case we really go to another page with a normal page unload and load.

Reply:
This is the piece of code that handles the navigation:
https://github.com/uyuni-project/uyuni/blob/master/web/html/src/core/spa/spa-engine.tsx#L107
and it does not trigger the `window.onbeforeunload` event.

Reply:
wait.... the console page is not loaded like the others I think... there is no SPA there since it's a separate full-page content
with no menu

Reply:
changed to `componentDidMount()` and `componentWillUnmount` anyway